### PR TITLE
[switch] Support torch.vmap with a batched index

### PR DESCRIPTION
**Pull request by bohnstingl**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #195326
* #195325
* #195324
* #195323
* #195322

Extends the switch vmap batch rule to a batched index, where every batch
element may take a different branch. A single switch cannot express that, so
this lowering evaluates all branches on the whole batch and selects their
outputs element-wise with `torch.where`. The index is truncated to int64 and
clamped into `[0, len(branches))` inside the selection function, matching what
the dense implementation and the `torch.switch` frontend do for an unbatched
index.

The two lowerings share `_vmap_with_front_bdims` from the previous commit: it
normalizes the out dims and strides of the vmapped branch outputs, which the
select path needs so that `torch.where` broadcasts against the index the same
way for every branch. `_check_matching_out_specs` is extracted from the two
existing callers now that the leaf-wise selection needs it as well.

Two behavioral consequences are documented in the docstring and in
docs/source/higher_order_ops/switch.md: every branch must be safe to evaluate
for every batch element (an unselected branch can still contribute inf/nan to
the gradients), and non-tensor leaves cannot be selected element-wise, so `int`
leaves must have the same value in every branch. Diverging `int` leaves, which
the unbatched path merges into an unbacked SymInt, raise instead.

An alternative would have been to sort or partition the batch by index and run
one switch per group, which avoids evaluating unselected branches but needs
data-dependent shapes, so it is not expressible under vmap. Selecting
element-wise keeps the transform shape-static at the cost of running every
branch.

The vmap tests are parametrized over `batched_index` so both lowerings are
covered by the same test bodies, plus tests for a nested switch on the same
batched index, a switch with no operands (only vmappable with a batched index,
since vmap needs at least one batched input), and the diverging-`int` error.

Test Plan:

```
python test/functorch/test_control_flow.py -k switch_vmap
python test/functorch/test_control_flow.py -k switch
python test/functorch/test_control_flow.py -k vmap
python test/inductor/test_control_flow.py -k switch
```

39 switch_vmap tests, 106 `-k switch`, 155 `-k vmap` and 18 inductor switch
tests pass. `lintrunner` is clean on the touched files (`spin` is not available
in this environment).

This change was authored with the help of an AI assistant (Claude Code).

Co-Authored-By: Claude Opus 5 <noreply@anthropic.com>

### CODE REVIEW

File: torch/_higher_order_ops/switch.py [hunk 1]
```diff
@@ -94,6 +94,15 @@ def _get_branch(branches, idx):
     return branches[idx]
 
 
+def _check_matching_out_specs(out_specs):
+    for i, spec in enumerate(out_specs):
+        if out_specs[0] != spec:
+            raise RuntimeError(
```
Comment (Skylion007) [inline torch/_higher_order_ops/switch.py]: Is this an AssertError or a ValueError?

**Comment by bohnstingl** [comment]
@claude please review these changes

**Comment by bohnstingl** [comment]
@ydwu4 This PR would be ready for your review
